File: mobile/modules/expo-litert-lm/src/ExpoLitertLmModule.ts
import { requireNativeModule } from 'expo-modules-core';
import type {
  TokenEventPayload,
  GenerationCompletePayload,
  GenerationErrorPayload,
} from './ExpoLitertLm.types';

// Native module interface — the module also exposes event listener methods
interface ExpoLitertLmNative {
  loadModel(modelPath: string): Promise<void>;
  generate(prompt: string): Promise<string>;
  generateStream(requestId: string, prompt: string): Promise<void>;
  unloadModel(): Promise<void>;
  isModelLoaded(): boolean;
  addListener(eventName: string, listener: (...args: any[]) => void): { remove: () => void };
  removeAllListeners(eventName: string): void;
}

// Cache native module reference
let cachedModule: ExpoLitertLmNative | null = null;
let moduleLoadFailed = false;

function getNativeModule(): ExpoLitertLmNative | null {
  if (moduleLoadFailed) return null;
  if (cachedModule) return cachedModule;
  try {
    cachedModule = requireNativeModule<ExpoLitertLmNative>('ExpoLitertLmModule');
    return cachedModule;
  } catch (error) {
    console.warn('[ExpoLitertLmModule] Native module not available:', error);
    moduleLoadFailed = true;
    return null;
  }
}

/**
 * Check if the native module is available on this device (Android only)
 */
export function isNativeModuleAvailable(): boolean {
  return getNativeModule() !== null;
}

/**
 * Load a LiteRT-LM model from the given path
 * This is a long-running operation (~10s). Resolves when the model is ready.
 *
 * @param modelPath - Absolute path to the .litertlm file (no file:// prefix)
 */
export async function nativeLoadModel(modelPath: string): Promise<void> {
  const mod = getNativeModule();
  if (!mod) throw new Error('LiteRT-LM native module not available');
  return mod.loadModel(modelPath);
}

/**
 * Generate a response for the given prompt (blocking)
 *
 * @param prompt - The prompt to generate from
 * @returns The full generated text
 */
export async function nativeGenerate(prompt: string): Promise<string> {
  const mod = getNativeModule();
  if (!mod) throw new Error('LiteRT-LM native module not available');
  return mod.generate(prompt);
}

/**
 * Generate a response with token streaming.
 * Resolves immediately; tokens are emitted via 'onToken' events.
 * Completion is signaled by 'onGenerationComplete' or 'onGenerationError'.
 *
 * @param requestId - Unique identifier for this request
 * @param prompt - The prompt to generate from
 */
export async function nativeGenerateStream(requestId: string, prompt: string): Promise<void> {
  const mod = getNativeModule();
  if (!mod) throw new Error('LiteRT-LM native module not available');
  return mod.generateStream(requestId, prompt);
}

/**
 * Unload the current model from memory
 */
export async function nativeUnloadModel(): Promise<void> {
  const mod = getNativeModule();
  if (!mod) throw new Error('LiteRT-LM native module not available');
  return mod.unloadModel();
}

/**
 * Check if a model is currently loaded (synchronous)
 */
export function nativeIsModelLoaded(): boolean {
  const mod = getNativeModule();
  if (!mod) return false;
  return mod.isModelLoaded();
}

/**
 * Subscribe to token events during streaming generation
 */
export function addTokenListener(
  listener: (event: TokenEventPayload) => void
): { remove: () => void } {
  const mod = getNativeModule();
  if (!mod) return { remove: () => {} };
  return mod.addListener('onToken', listener);
}

/**
 * Subscribe to generation completion events
 */
export function addGenerationCompleteListener(
  listener: (event: GenerationCompletePayload) => void
): { remove: () => void } {
  const mod = getNativeModule();
  if (!mod) return { remove: () => {} };
  return mod.addListener('onGenerationComplete', listener);
}

/**
 * Subscribe to generation error events
 */
export function addGenerationErrorListener(
  listener: (event: GenerationErrorPayload) => void
): { remove: () => void } {
  const mod = getNativeModule();
  if (!mod) return { remove: () => {} };
  return mod.addListener('onGenerationError', listener);
}

File: mobile/modules/expo-litert-lm/android/src/main/java/expo/modules/litertlm/ExpoLitertLmModule.kt
package expo.modules.litertlm

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import com.google.ai.edge.litertlm.Engine
import com.google.ai.edge.litertlm.EngineConfig
import com.google.ai.edge.litertlm.Message
import com.google.ai.edge.litertlm.Content

/**
 * ExpoLitertLmModule — Native LiteRT-LM inference for Expo (Android only)
 *
 * API mapping:
 *   Message.of(text: String) → user Message
 *   sendMessageAsync(Message) → Flow<Message>   (partial/delta tokens)
 *   Content.Text.text → extracted token text
 *   Conversation implements AutoCloseable (.use {} supported)
 */
class ExpoLitertLmModule : Module() {
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
  private var engine: Engine? = null
  private var modelPath: String? = null

  override fun definition() = ModuleDefinition {
    Name("ExpoLitertLmModule")

    Events("onToken", "onGenerationComplete", "onGenerationError")

    /**
     * Load a LiteRT-LM model from the given path.
     * This is a long-running operation (~10s). Resolves when ready.
     */
    AsyncFunction("loadModel") { path: String, promise: Promise ->
      scope.launch {
        try {
          android.util.Log.d("ExpoLitertLm", "Loading model: $path")

          // Unload existing engine if any
          engine?.close()
          engine = null

          val config = EngineConfig(path)
          val newEngine = Engine(config)
          newEngine.initialize()

          engine = newEngine
          modelPath = path

          android.util.Log.d("ExpoLitertLm", "Model loaded successfully: $path")
          promise.resolve(null)
        } catch (e: Exception) {
          android.util.Log.e("ExpoLitertLm", "Failed to load model: ${e.message}")
          promise.reject("LOAD_ERROR", "Failed to load LiteRT-LM model: ${e.message}", e)
        }
      }
    }

    /**
     * Generate a response for the given prompt (blocking until complete).
     * Returns the full generated text.
     */
    AsyncFunction("generate") { prompt: String, promise: Promise ->
      scope.launch {
        val currentEngine = engine
        if (currentEngine == null) {
          promise.reject("NOT_LOADED", "No model loaded. Call loadModel() first.", null)
          return@launch
        }

        try {
          android.util.Log.d("ExpoLitertLm", "Generating response for prompt length: ${prompt.length}")
          val sb = StringBuilder()
          val userMessage = Message.of(prompt)

          currentEngine.createConversation().use { conversation ->
            conversation.sendMessageAsync(userMessage).collect { partialMessage ->
              val tokenText = partialMessage.contents
                .filterIsInstance<Content.Text>()
                .joinToString("") { it.text }
              sb.append(tokenText)
            }
          }

          val result = sb.toString()
          android.util.Log.d("ExpoLitertLm", "Generation complete. Output length: ${result.length}")
          promise.resolve(result)
        } catch (e: Exception) {
          android.util.Log.e("ExpoLitertLm", "Generation failed: ${e.message}")
          promise.reject("GENERATE_ERROR", "LiteRT-LM generation failed: ${e.message}", e)
        }
      }
    }

    /**
     * Generate a response with streaming.
     * Resolves immediately; tokens are emitted via 'onToken' events.
     * Signals completion via 'onGenerationComplete' or 'onGenerationError'.
     */
    AsyncFunction("generateStream") { requestId: String, prompt: String, promise: Promise ->
      val currentEngine = engine
      if (currentEngine == null) {
        promise.reject("NOT_LOADED", "No model loaded. Call loadModel() first.", null)
        return@AsyncFunction
      }

      // Resolve the promise immediately — streaming happens via events
      promise.resolve(null)

      scope.launch {
        try {
          android.util.Log.d("ExpoLitertLm", "Starting stream for requestId: $requestId")
          val sb = StringBuilder()
          val userMessage = Message.of(prompt)

          currentEngine.createConversation().use { conversation ->
            conversation.sendMessageAsync(userMessage).collect { partialMessage ->
              val tokenText = partialMessage.contents
                .filterIsInstance<Content.Text>()
                .joinToString("") { it.text }
              if (tokenText.isNotEmpty()) {
                sb.append(tokenText)
                sendEvent("onToken", mapOf(
                  "requestId" to requestId,
                  "token" to tokenText
                ))
              }
            }
          }

          val fullText = sb.toString()
          android.util.Log.d("ExpoLitertLm", "Stream complete for requestId: $requestId, chars: ${fullText.length}")
          sendEvent("onGenerationComplete", mapOf(
            "requestId" to requestId,
            "fullText" to fullText,
            "tokenCount" to fullText.length
          ))
        } catch (e: Exception) {
          android.util.Log.e("ExpoLitertLm", "Stream error for requestId: $requestId — ${e.message}")
          sendEvent("onGenerationError", mapOf(
            "requestId" to requestId,
            "error" to (e.message ?: "Unknown error")
          ))
        }
      }
    }

    /**
     * Unload the current model from memory
     */
    AsyncFunction("unloadModel") { promise: Promise ->
      scope.launch {
        try {
          engine?.close()
          engine = null
          modelPath = null
          android.util.Log.d("ExpoLitertLm", "Model unloaded")
          promise.resolve(null)
        } catch (e: Exception) {
          android.util.Log.e("ExpoLitertLm", "Unload error: ${e.message}")
          promise.reject("UNLOAD_ERROR", "Failed to unload model: ${e.message}", e)
        }
      }
    }

    /**
     * Check if a model is currently loaded (synchronous)
     */
    Function("isModelLoaded") {
      engine != null
    }

    OnDestroy {
      scope.cancel()
      engine?.close()
      engine = null
    }
  }
}
